Support for searching for previous event in the Events Table

bhufmann opened the pull request:
- Add a 'previous' button in the column search cells right beside the
'next' button
- Use keyboard short-cut SHIFT+ENTER to search previous event

fixes #412

Signed-off-by: Bernd Hufmann <Bernd.Hufmann@ericsson.com>

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -421,6 +428,9 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
                 ['isFiltered']: isFiltered
             };
+            if (direction !== undefined) {
+                additionalParams['table_search_direction'] = direction.toString;
```
This actually puts the toString function reference in the map ;) it's missing () after toString...

bhufmann replied:
good catch

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -490,11 +500,11 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
     }
 
-    private async findNextMatchIndex(currRowIndex: number) {
+    private async findMatchIndex(currRowIndex: number, direction = Direction.NEXT) {
```
This isn't working as expected... Direction.NEXT is actually the number 0 and Direction.PREVIOUS is the number 1. So later the toString() in the map puts "0" or "1"...

bhufmann replied:
Direction[direction] seems to return the string

bhufmann commented:
@muddana-satish could you please review this change? The latest nightly build has the corresponding back-end code implementation.

https://download.eclipse.org/tracecompass.incubator/trace-server/rcp/

muddana-satish commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -528,7 +538,42 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
+    }
+
+    private async findPreviousMatchedEvent() {
+        let isFound = false;
+        if (this.gridApi) {
+            this.gridApi.deselectAll();
+            const rowNodes: RowNode[] = [];
+
+            const currRowIndex = this.selectStartIndex === -1 ? this.gridApi?.getLastDisplayedRow() + 1 : this.selectStartIndex;
+            this.gridApi.forEachNode(rowNode => rowNodes.push(rowNode));
+
+            rowNodes.reverse().forEach(rowNode => {
+                if (rowNode.rowIndex < Math.max(this.selectStartIndex, this.selectEndIndex) && rowNode.data
+                    && rowNode.data['isMatched'] && !isFound) {
+                    this.gridApi?.ensureIndexVisible(rowNode.rowIndex);
+                    this.selectStartIndex = this.selectEndIndex = rowNode.rowIndex;
+                    this.handleRowSelectionChange();
+                    rowNode.setSelected(true);
+                    isFound = true;
+                }
+            });
+
+            if (isFound) {
+                return;
+            }
+
```
findPreviousMatchedEvent and findNextMatchedEvent share a lot of things. Can we reuse the logic?

bhufmann replied:
I'll keep the tags API because there might be other client implementations than the theia-trace-extension that relies on the server side filtering. 

I'll remove the code for 'isFiltered' flag since the server is updated. I'll also try to minimize the code duplication findPreviousMatchedEvent and findNextMatchedEvent.

muddana-satish reviewed (dismissed):
Looks good to me.

I don't know if we want to fix it in this but we no longer need the 'isFiltered' flag. Also we don't need additional information from the api like 'tags'.

bhufmann commented:
I updated the patch according to the review comments. 

I also noticed a general bug with the searching. The AgGrid keeps a cache of size maxBlocksInCache*cacheBlockSize (in our case (5*200)=1000), where each block of cacheBlockSize events are fetched from the back-end. These 5 blocks, however, are not necessarily consecutive blocks, it depends on user's navigation. In the search feature, where we do this.gridApi.forEachNode(...) to move to the next match, the whole cache is looped through. It could happen that the first matched event in the cache is not inside the current block and the table moves somewhere else in the table where the next matched event in the cache is found. I corrected that in this PR as well.

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -508,13 +516,28 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
             this.gridApi.deselectAll();
+            let rowNodes: RowNode[] = [];
+            let currRowIndex = this.selectStartIndex === -1 ? this.gridApi?.getLastDisplayedRow() + 1 : this.selectStartIndex;
+            const blockSize = this.props.cacheBlockSize ? this.props.cacheBlockSize : 200;
```
Just checking, will the 'current block' fix work correctly if the cachBlockSize is undefined and we use 200?

bhufmann replied:
I just chose 200 because that's the current blocksize. I don't think there is a "good" value if the cacheBlockSize is undefined.

PatrickTasse replied:
To clarify, what I'm worried is if we accurately compute the 'block number' for comparison in that case.

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -508,13 +516,28 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
+            rowNodes.forEach(rowNode => {
+                if ((rowNode.data && rowNode.data['isMatched'] && !isFound) &&
+                    (((direction === Direction.PREVIOUS) &&
+                        (rowNode.rowIndex < Math.max(this.selectStartIndex, this.selectEndIndex))) ||
```
Will this prevent searching backwards if there is no selection (both are -1, right?)

bhufmann replied:
mhh, maybe it should compare with currRowIndex instead? I'll check

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -526,9 +549,11 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
-            const currRowIndex = this.gridApi?.getLastDisplayedRow() + 1;
-            const lineIndex = await this.findNextMatchIndex(currRowIndex);
+            if (direction === Direction.NEXT) {
+                currRowIndex = this.gridApi?.getLastDisplayedRow() + 1;
```
I'm not sure if necessary, but maybe it needs to be the first index following the last row node index that was considered by the 'current block' fix above?

And for PREVIOUS it should be the first index before the first considered row node index?

bhufmann replied:
I'll check

PatrickTasse reviewed (changes requested):
Some more suggestions perhaps for future enhancement:

- It should be possible to use Enter/Shift+Enter when the table itself has focus
- There should be a visual feedback while we are searching the next/previous index, not only when we are waiting to fill the table at the found index
- It should not lose current table selection if there is no found index

bhufmann commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -508,13 +516,28 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
         if (this.gridApi) {
             this.gridApi.deselectAll();
+            let rowNodes: RowNode[] = [];
+            let currRowIndex = this.selectStartIndex === -1 ? this.gridApi?.getLastDisplayedRow() + 1 : this.selectStartIndex;
```
I'll check if this works

PatrickTasse replied:
If there is no selection I would expect the search would find the first match in the visible rows if there is any?

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -508,13 +516,59 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
+            rowNodes.forEach(rowNode => {
+                let condition = this.selectStartIndex === -1 ? currentBlock * blockSize : Math.max(this.selectStartIndex, this.selectEndIndex);
+                if (direction === Direction.PREVIOUS) {
+                    condition = this.selectStartIndex === -1 ? (currentBlock + 1) * blockSize - 1 : Math.min(this.selectStartIndex, this.selectEndIndex);
```
I propose that forward search start at index 0 when there is no selection, and backward search be a noop when there is no selection.

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -508,13 +516,59 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
-                if (rowNode.rowIndex > Math.max(this.selectStartIndex, this.selectEndIndex) && rowNode.data
-                    && rowNode.data['isMatched'] && !isFound) {
+                // make sure to search only in current block and not in full cache
+                if (currentBlock === Math.floor(rowNode.rowIndex / blockSize)) {
```
This prevents considering contiguous blocks. I propose to use a scheme that only considers contiguous row index regardless of their block.

PatrickTasse commented on packages/react-components/src/components/table-output-component.tsx:
```diff
@@ -524,14 +578,19 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
-            const lineIndex = await this.findNextMatchIndex(currRowIndex);
-            if (lineIndex) {
+            // find match outside the current block
+            const lineIndex = await this.findMatchIndex(currRowIndex, direction);
```
This sends the index of a row that should not be searched, then the server skips that index and starts at the next/previous index.
I propose that the client should send the first index that should be searched (so after/before the selection), and the server be changed to start at that index. See https://git.eclipse.org/r/c/tracecompass/org.eclipse.tracecompass/+/183874.
